File: backend/services/revenue_service.py
import logging
from apps.Revenue.backend.database import RevenueRepository
from apps.Revenue.backend.sanitizer import build_key_map, sanitize

logger = logging.getLogger(__name__)


class RevenueService:
    @staticmethod
    async def get_clean_revenue() -> list:
        """Fetches all revenue rows and sanitizes them."""
        raw_rows = RevenueRepository.find_all()
        if not raw_rows:
            return []
        key_map = build_key_map(raw_rows[0])
        clean_data = []
        for row in raw_rows:
            sanitized = sanitize(row, key_map)
            if sanitized and sanitized.get('date'):
                # Convert date to ISO string for JSON
                sanitized['date'] = sanitized['date'].isoformat()
                clean_data.append(sanitized)
        return clean_data

File: apps/api/services/revenueService.js
import { RevenueRepository } from '../repositories/revenueRepository.js';
import { DataLogic } from '../../../packages/shared/dist/index.js';

/**
 * SERVICE LAYER: Orchestrates Repository data and Shared Domain logic.
 */
export class RevenueService {
    /**
     * Retrieves all revenue records and sanitizes them using the isomorphic DataLogic engine.
     * Ensures parity with frontend data expectations.
     */
    static async getCleanRevenue() {
        const rawRows = await RevenueRepository.findAll();
        if (!rawRows || rawRows.length === 0) return [];

        // Build key map once based on the first row (MIT Engineering: efficiency)
        const keyMap = DataLogic.buildKeyMap(rawRows[0]);
        const cleanData = [];

        rawRows.forEach(row => {
            const sanitized = DataLogic.sanitize(row, keyMap);
            if (sanitized && !isNaN(sanitized.date.getTime())) {
                // Flatten for API delivery - ensure date is ISO string for JSON
                cleanData.push({
                    ...sanitized,
                    date: sanitized.date.toISOString()
                });
            }
        });

        return cleanData;
    }
}
